src/Deserialization.cpp

```diff
@@ -0,0 +1,1358 @@
+    assert(exprs_types != nullptr);
+    assert(exprs_serialized != nullptr);
+    std::vector<Expr> result;
+    result.reserve(exprs_serialized->size());
+    for (size_t i = 0; i < exprs_serialized->size(); ++i) {
+        auto expr = deserialize_expr(static_cast<Serialize::Expr>(exprs_types->Get(i)), exprs_serialized->Get(i));
+        result.push_back(expr);
+    }
+    return result;
+}
+
+Range Deserializer::deserialize_range(const Serialize::Range *range) {
+    assert(range != nullptr);
+    auto min = deserialize_expr(range->min_type(), range->min());
+    auto extent = deserialize_expr(range->extent_type(), range->extent());
+    return Range(min, extent);
+}
+
+Bound Deserializer::deserialize_bound(const Serialize::Bound *bound) {
+    assert(bound != nullptr);
+    auto var = deserialize_string(bound->var());
+    auto min = deserialize_expr(bound->min_type(), bound->min());
+    auto extent = deserialize_expr(bound->extent_type(), bound->extent());
+    auto modulus = deserialize_expr(bound->modulus_type(), bound->modulus());
+    auto remainder = deserialize_expr(bound->remainder_type(), bound->remainder());
+    auto hl_bound = Bound();
+    hl_bound.var = var;
+    hl_bound.min = min;
+    hl_bound.extent = extent;
+    hl_bound.modulus = modulus;
+    hl_bound.remainder = remainder;
+    return hl_bound;
+}
+
+StorageDim Deserializer::deserialize_storage_dim(const Serialize::StorageDim *storage_dim) {
+    assert(storage_dim != nullptr);
+    auto var = deserialize_string(storage_dim->var());
+    auto alignment = deserialize_expr(storage_dim->alignment_type(), storage_dim->alignment());
+    auto bound = deserialize_expr(storage_dim->bound_type(), storage_dim->bound());
+    auto fold_factor = deserialize_expr(storage_dim->fold_factor_type(), storage_dim->fold_factor());
+    auto fold_forward = storage_dim->fold_forward();
+    auto hl_storage_dim = StorageDim();
+    hl_storage_dim.var = var;
+    hl_storage_dim.alignment = alignment;
+    hl_storage_dim.bound = bound;
+    hl_storage_dim.fold_factor = fold_factor;
+    hl_storage_dim.fold_forward = fold_forward;
+    return hl_storage_dim;
+}
+
+LoopLevel Deserializer::deserialize_loop_level(const Serialize::LoopLevel *loop_level) {
+    assert(loop_level != nullptr);
+    auto func_name = deserialize_string(loop_level->func_name());
+    auto stage_index = loop_level->stage_index();
+    auto var_name = deserialize_string(loop_level->var_name());
+    auto is_rvar = loop_level->is_rvar();
+    auto locked = loop_level->locked();
+    return LoopLevel(func_name, var_name, is_rvar, stage_index, locked);
+}
+
+FuncSchedule Deserializer::deserialize_func_schedule(const Serialize::FuncSchedule *func_schedule) {
+    assert(func_schedule != nullptr);
+    auto store_level = deserialize_loop_level(func_schedule->store_level());
+    auto compute_level = deserialize_loop_level(func_schedule->compute_level());
+    std::vector<StorageDim> storage_dims;
```

Comment:
The pattern of

     declare vector
     reserve vector
     populate vector by calling some fn on every instance in another vector

occurs often enough here that I wonder if it would be useful to make a templated helper func to make it terser (and to ensure that reserve() is always called). What do you think? eg

```
template<typename Src, typename Dst>
std::vector<Dst> deserialize_vector(flatbuffer_collection<Src>, std::function<Dst(Src)>);
```

Reply:
I like this!